### ISSUE TITLE
[bug] MReSC generates wrapper with same seeds for LFSRs

### ISSUE THREAD

**Issue by danilo-bc** [arminalaghi/scsynth#4]
It has been confirmed that MReSC and ReSC have different LFSR generators. While VerilogSCGenerator has 3 possibilities, VerilogMultivariateReSCGenerator only has one, the SharedLFSR method. (A small patch shall be merged soon about this)

The main problem is the wrappers. 

The single variable VerilogSCWrapperGenerator.m (line 131 and beyond) succesfully generates instances of LFSRs with different seeds for Inputs and Constants, also taking in consideration 'degrees'.

The multivariable VerilogMultivariateSCWrapperGenerator.m (line 101 and beyond), on the other hand, doesn't take in consideration number of variables. Thus, it generates LFSRs with the same seeds for different inputs, which will change the stochastic function synthesized!

On line 106 of VerilogMultivariateSCWrapperGenerator.m:
```octave
fprintf(fp, '\t\t.seed (%d''d%d),\n', m, round(N*j/(degrees(i)*2+1)));
```

This expression should, somehow, incorpate the value "i" (the loop count variable) to solve this problem.

Since I'm not currently familiar with the reasoning behind the value `round(N*j/(degrees(i)*2+1))`, I don't know how to properly modify it.

**Comment by danilo-bc** [comment]
I still didn't have time to properly look into this, but I wanted to post again to remind this is an issue.

**Comment by arminalaghi** [comment]
Apologies for my extra long delay.
You are right. We need to uniquify the seed. What do you think about replacing Line 106 with this?
`round(N*(0.5*(i+j)*(i+j+1)+j)/(degrees(i)*2+1))`

Basically changin `N*j` to `N*(0.5*(i+j)*(i+j+1)+j)`. This is the [Cantor pairing function](https://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function).

Let me know what you think.

**Comment by danilo-bc** [comment]
No apologies needed, I'm a bit busy myself, these last days.

I haven't delved into this solution yet, because what bothers me is the lack of SCC analysis. If the inputs are certainly independent from each other (or at least x1 has independent copies of itself if degree>1), shouldn't it be guaranteed that SCC = 0?

Am I missing some sort of shortcut where using certain building blocks already guarantees SCC = 0 without need of further analysis?

**Comment by danilo-bc** [comment]
Since I had a bit of time, I implemented the SCC(X,Y) function from your article about exploiting correlation to help synthesize circuits.

The problem is SCC is supposed to yield only -1,+1 and 0, but certain pairs of streams give intermediate values.

For instance, take X = 10010100 and Y = 00010011. This gives SCC(X,Y) = -0.11111...

Another problem that arises is that shifting X and Y with the LFSR formula provided (XOR with taps on bits 0, 1, 2 and 7, in this case) has the SCC values altering from +1 to 0 to -0.1111..., 0.3333... and such.

**EDIT:** on second thought, SCC of the LFSRs themselves is not important, but the SCC of the stochastic numbers they generate. Is there a procedural way to guarantee that, given a few LFSRs, SNs will have SCC = 0? (other than creating a testbench for a few binary numbers to turn into SNs and then calculate all their SCCs)

**Comment by arminalaghi** [comment]
This is a great discussion point, and has been a topic of research for a while now.

It is possible to guarantee SCC=0 using LFSRs or plain binary counters. This paper shows how it is done: https://eprints.soton.ac.uk/259810/1/GeneratingBinarySequencesForStochasticComputing.pdf
(quick summary: if you add an all-zero state to an m-sequence LFSR and take it through its full period, the bit-streams generated by each LFSR bit will all have value 0.5 and will have SCC=0. The same applies to a binary counter; all its bits will have SCC=0 if you run it through its full period. To generate arbitrary SNs with SCC=0, you'll use bits of the LFSR/counter and feed them to the comparator.)

This method works well if you only want two numbers with SCC=0, but becomes impractical for more than 3 inputs. Basically, to generate SNs for k m-bit numbers, you will need 2^(k.m). The k factor in the exponent makes this approach impractical. And unfortunately, this case applies to ReSC and STRAUSS circuits. Depending on the degree of the polynomial, we may need multiple SNs at the input, and guaranteeing SCC=0 on them requires long bit-stream sequences.

Which takes us to the next best thing: LFSRs with shifted values and/or different taps/seeds. As you experiences, this arrangement of LFSRs does not necessarily produce numbers with SCC=0, but it gets us pretty close, and that's what many people have been using. There has been some studies on selecting optimal seeds for the LFSRs: https://ieeexplore.ieee.org/document/7459560

Low-discrepancy sequences are also good alternatives, but still cannot guarantee SCC=0. And they too become expensive when multiple inputs are needed.

For the current version of ReSC and STRAUSS, we basically select some random seeds for the LFSRs and hope for the best. That being said, the bug still needs to be fixed. Did you get a chance to try it with my suggested seed? Do the results improve?

**Comment by danilo-bc** [comment]
Thank you for the elaborate answer! I realized "using random seeds and praying for the best" was a valid option, but that wasn't too scientific to call to discussion, hehe.

I haven't tried the method yet, but I'll experiment, let the program compile and report back (which may take a bit of time because I'm still a bit tight on schedule).

My case study has been a 3x3 Sobel Filter for edge detection, which sounds minimal for any deterministic implementation, but rather challenging when the need for 8+ uncorrelated inputs becomes evident.

I'll read the references mentioned, test the Cantor pairing function and, hopefully, have a good enough error margin for so many inputs!

**EDIT:** results are promising. Since the code takes a rather long time to run for my case study, I haven't opened a pull request yet. Given that there's currently no need (or formal way) of guaranteeing a certain degree of correlation between inputs, the problem seeems solved.

**Comment by danilo-bc** [comment]
In hindsight, although the function gave different values for all the seeds, the values were invalid.

I took a look at this simple, practical Python implementation of the Cantor Pairing Function and its inverse (https://github.com/perrygeo/pairing) and my final methodology is:

1. Apply pairing function `temp = int(0.5 * (i+j) * (i+j+1) + j)` for i being the number of input variables (size of degrees) and j being that specific variable's degree (degrees(i))
2. Apply normalization from 0 to 2^(m-1) or N-1 (stochastic stream size) `seeds = (seeds - min(seeds))*((N-1)-0)/(max(seeds) - min(seeds)) - 0` which, ommiting the 0's from regular linear normalization becomes: `seeds = (seeds - min(seeds))*(N-1)/(max(seeds) - min(seeds))`. The last step is applying the rounding function and making sure every number is an integer after scaling `seeds = int(round(seeds))`

I tested this on python, for simplicity, and I could map 90 values to 90 unique seed values from 0 to 256, which is quite an extreme case. This meant 15 input variables with degree 6 for each. When I tried the same function for N = 128, it could only find 62 unique values. I identify that more as a project flaw or tool limitation, since there's only so much that can be automated. Besides, finding about 90 uncorrelated values in a 128 number space seems quite unreasonable for 7 or 8-bit LFSRs, so I propose leaving that to another discussion.

I'll modify the appropriate portion and open a new pull request.

**MINOR EDIT:** since this paring and scaling produces both 0 and N-1 (which are 1 step away from each other in all LFSR implementations), I will scale up to N-2, instead. This, as far as I know, doesn't necessarily relate to 0 in such a direct way. If it interacts with other seeds in the scaling process in any way, the analysis becomes picking optimal seeds, which, by itself, is a broad area of study.
